src/searchModal/advancedSearch.js

```diff
@@ -0,0 +1,453 @@
+import 'ui/datatable';
+import 'select2';
+import urlUtil from 'util/url';
+import request from 'core/dataProvider/request';
+
+/**
+ * Creates advanced search component
+ *
+ * @param {object} config
+ * @param {object} config.renderTo - DOM element where component will be rendered to
+ * @param {string} config.advancedCriteria - advanced criteria to be set on component creation
+ * @returns {advancedSearch}
+ */
+export default function advancedSearchFactory(config) {
+    // Private properties to be easily accessible by instance methods
+    let $container = null;
+    let $addCriteriaInput = null;
+    let $criteriaSelect = null;
+    let $advancedCriteriaContainer = null;
+    let criteriaState = null;
+
+    // Creates new component
+    const instance = component({
+        /**
+         * Request metadata (criteria) for the given uri
+         * @param {string} classUri - Uri of the class to get metadata from
+         * @param {number} maxListSize - max number of values to return for a metadata of type list If
+         *                               available values options exceed this number, an uri will
+         *                               be returned instead to get its values dinamically
+         * @returns {Promise} - Request promise
+         */
+        updateCriteria: function (classUri, maxListSize = 5) {
+            const route = urlUtil.route('get', 'ClassMetadata', 'tao', { classUri, maxListSize });
+            return request(route)
+                .then(response => {
+                    const criteria = formatCriteria(response);
+                    updateCriteria(criteria);
+                })
+                .catch(e => instance.trigger('error', e));
+        },
+        /**
+         * Access to component state
+         * @returns {Object} - criteria state
+         */
+        getState: function () {
+            return criteriaState;
+        },
+        /**
+         * Removes every rendered criterion, updates criteria state accordingly
+         * and removes classes applied to scrollable list of criteria
+         */
+        clear: function () {
+            $advancedCriteriaContainer.removeClass('scrollable');
+            $advancedCriteriaContainer.removeClass('scroll-separator-top');
+            $advancedCriteriaContainer.removeClass('scroll-separator-bottom');
+            $advancedCriteriaContainer.empty();
+            _.forEach(criteriaState, criterion => {
+                criterion.rendered = false;
+                criterion.value = undefined;
+            });
+        },
+        /**
+         * Builds substring of search query with the advanced criteria conditions
+         */
+        getAdvancedCriteriaQuery: function () {
+            const advancedSearchCriteria = _.filter(criteriaState, criterion => criterion.rendered === true);
+            let query = '';
+
+            advancedSearchCriteria.forEach(renderedCriterion => {
+                if (renderedCriterion.type === 'text') {
```

Comment:
Yes, I considered it in case there were many types of criteria, although as finally we only defined two, I did not implemented an enum. But it is true that this two strings are used in different places and will become more complex in the future, so I have updated it and included an  enum to manage it.